### PR TITLE
ohos: Add basic IME and keyboard support

### PR DESCRIPTION
**Pull request by jschwe**
- Add extremely basic support for keyboard events
- Add basic IME support
   - Showing and hiding the IME
   - inserting text
   - deleting characters (implemented by simulating backspace / forward delete key presses)
   - very basic configuration of the IME

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

### CODE REVIEW

**Comment by wusyong** [comment]
I wonder how Ohos detects if it's on IME. Will it always be input with IME?
I'm thinking about how desktop's winit event can map to it: https://docs.rs/winit/latest/winit/event/enum.Ime.html

**Comment by jschwe** [comment]
The xcomponent keyevent callback API is for physical keyboards. I haven't look into that very deeply though, since on mobile basically everything goes through the IME (soft keyboard == IME).

Regarding winit, There are two callbacks `set_preview_text()` and `finish_preview_text()` which should map well to the pre-edit and commit winit events (but which [I haven't implemented yet](https://github.com/openharmony-rs/ohos-ime/blob/72472ebe9ed17a00dba3d9b4731cf4d36197fd4b/src/text_editor.rs#L328-L339), since it seems that text input in Chinese works just fine without implementing these callbacks)

**Comment by wusyong** [comment]
I see. My experience is that it will depends on platforms. Windows will need some checks while macOS can input some kanji already. I'll check again how to map to them in the future.

**Comment by jschwe** [comment]
> I see. My experience is that it will depends on platforms.

just to clarify: openharmony is an independent operating system.

**Review by jdm — APPROVED** [review]

File: ports/servoshell/Cargo.toml [hunk 1]
```diff
@@ -91,6 +92,9 @@ napi-derive-ohos = "1.0.1"
 napi-ohos = "1.0.1"
 ohos-sys = { version = "0.4.0", features = ["xcomponent"] }
 ohos-vsync = "0.1.2"
+ohos-ime = { version = "0.2" }
+ohos-ime-sys = "0.1.1"
+xcomponent-sys = { version = "0.1.1", features = ["api-12", "keyboard-types"] }
```
Comment (mukilan) [inline ports/servoshell/Cargo.toml:97]: Does the `api-12` here imply we need to build against the OHOS 5.0 SDK?

**Review by mukilan — COMMENTED** [review]
Have we run `cargo-clippy` on this code? The CI currently only checks the linux code, and it would be good to avoid new warnings here.

File: ports/servoshell/Cargo.toml (hunk 1 above)
Comment (jschwe) [reply ports/servoshell/Cargo.toml:97]: Building works fine with the OH 4.1 SDK, but at runtime OH 5.0 is required. The whole IME interface for native code was added in API-12.

File: components/shared/compositing/constellation_msg.rs [hunk 2]
```diff
@@ -34,6 +34,8 @@ pub enum ConstellationMsg {
     IsReadyToSaveImage(HashMap<PipelineId, Epoch>),
     /// Inform the constellation of a key event.
     Keyboard(KeyboardEvent),
+    /// Inform the constellation of a composition event (IME).
+    CompositionEvent(CompositionEvent),
```
[the remark this answers is not in the document]
Comment (jschwe) [reply components/shared/compositing/constellation_msg.rs:38]: Sounds reasonable. I think I just chose the same name as the inner `CompositionEvent`.

File: ports/servoshell/Cargo.toml (hunk 1 above)
Comment (mukilan) [reply ports/servoshell/Cargo.toml:97]: I get this error when building this branch with SDK version 4.1.7.5:

```
  = note: ld.lld: error: unable to find library -lohinputmethod
          clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
```

Maybe I'm missing something in the configuration?

Comment (jschwe) [reply ports/servoshell/Cargo.toml:97]: Huh, I could have sworn I was still using the 4.1 SDK on my mac. I guess then we need to bump the SDK version in CI. It seems the book doesn't mention the minimum supported API version right now, but probably we should mention that too.

File: ports/servoshell/egl/ohos.rs [hunk 3]
```diff
@@ -143,6 +160,26 @@ impl ServoAction {
...
+                Ok(())
+            },
+            ImeDeleteBackward(len) => {
+                for _ in 0..*len {
+                    let _ = servo.key_down(Key::Backspace);
+                    let _ = servo.key_up(Key::Backspace);
+                }
+                Ok(())
+            },
+            ImeSendEnter => servo
```
[the remark this answers is not in the document]
Comment (jschwe) [reply ports/servoshell/egl/ohos.rs:180]: Adding a newline would be possible though.

**Comment by jschwe** [comment]
> Have we run cargo-clippy on this code? The CI currently only checks the linux code, and it would be good to avoid new warnings here.

I checked locally and this doesn't seem to add any new warnings. I fixed some existing ones, but probably it would make sense to make a follow-up clippy PR and add a CI job

File: ports/servoshell/egl/ohos.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jschwe) [reply ports/servoshell/egl/ohos.rs:180]: newline added

File: ports/servoshell/Cargo.toml (hunk 1 above)
Comment (jschwe) [reply ports/servoshell/Cargo.toml:97]: I bumped the version to 5.0 in the arkts project and CI

Comment (mukilan) [reply ports/servoshell/Cargo.toml:97]: > It seems the book doesn't mention the minimum supported API version right now, but probably we should mention that too.

Yep, the book needs to be updated. We also enforce that in mach, so we should update [this](https://github.com/servo/servo/blob/aa7116c75d2f13f554ce4db162344aaea59e60b0/python/servo/platform/build_target.py#L278) in this PR.

Comment (jschwe) [reply ports/servoshell/Cargo.toml:97]: Thanks! I also updated the check in `mach` and upgraded it from a warning to an error if the SDK is too old.

**Review by mukilan — APPROVED** [review]

**Pull request merged**